cubeb-ffi/src/ffi.rs

```diff
@@ -0,0 +1,674 @@
+use libc::{c_long,c_void};
+use std::ops;
+use std::default::Default;
+
+#[macro_export]
+macro_rules! log_internal {
+    ($level: expr, $msg: expr) => {
+        #[allow(unused_unsafe)]
+        unsafe {
+            if $level <= $crate::g_log_level {
+                if let Some(log_callback) = $crate::g_log_callback {
+                    let cstr = ::std::ffi::CString::new(concat!("%s:%d: ", $msg, "\n")).unwrap();
+                    log_callback(cstr.as_ptr(), file!(), line!());
+                }
+            }
+        }
+    };
+    ($level: expr, $fmt: expr, $($arg:tt)+) => {
+        #[allow(unused_unsafe)]
+        unsafe {
+            if $level <= $crate::g_log_level {
+                if let Some(log_callback) = $crate::g_log_callback {
+                    let cstr = ::std::ffi::CString::new(concat!("%s:%d: ", $fmt, "\n")).unwrap();
+                    log_callback(cstr.as_ptr(), file!(), line!(), $($arg)+);
+                }
+            }
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! logv {
+    ($msg: expr) => (log_internal!($crate::LogLevel::Verbose, $msg));
+    ($fmt: expr, $($arg: tt)+) => (log_internal!($create::LogLevel::Verbose, $fmt, $($arg)*));
+}
+
+#[macro_export]
+macro_rules! log {
+    ($msg: expr) => (log_internal!($crate::LogLevel::Normal, $msg));
+    ($fmt: expr, $($arg: tt)+) => (log_internal!($crate::LogLevel::Normal, $fmt, $($arg)*));
+}
+
+//     g_log_callback("%s:%d: " fmt "\n", __FILE__, __LINE__, ##__VA_ARGS__); \
+
+pub enum Context {}
+pub enum Stream {}
+
+// TODO endian check
+pub const SAMPLE_S16NE: SampleFormat = SampleFormat::S16LE;
+pub const SAMPLE_FLOAT32NE: SampleFormat = SampleFormat::Float32LE;
```

Comment:
Ah, it wasn't clear these were part of cubeb_sample_format since the other members are part of enum SampleFormat.  It's a bit weird this way since they have such different naming, but I guess it's okay for now.